screensaverd.c

```diff
@@ -0,0 +1,220 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#define _BSD_SOURCE
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <dirent.h>
+#include <signal.h>
+#include <stdarg.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <sys/select.h>
+
+#define UNUSED			__attribute__((unused))
+
+#define DISPLAY_CONTROL		"/sys/class/graphics/fb0/blank"
+#define DEV_INPUT_DIR		"/dev/input"
+#define LOG_NAME		"screensaverd"
+#define EVENT_PREFIX		"event"
+#define MAX_DEVICES		32
+#define DISPLAY_OFF_TIME	25 /* seconds */
+#define BUF_SIZE		4096
+
+sig_atomic_t volatile running = 1;
+
+/* ------------------------------------------------------------------------ */
+
+/* perror() with printf-style format and application name prefix. */
+static void
+perrorf(const char *format, ...)
+{
+	va_list ap;
+
+	va_start(ap, format);
+	printf("[%s] ", LOG_NAME);
+	vprintf(format, ap);
+	printf(": %s\n", strerror(errno));
+	va_end(ap);
+}
+
+/* ------------------------------------------------------------------------ */
+
+static void
+close_fds(int fds[], int num)
+{
+	int i;
+
+	for (i = 0; i < num; i++)
+		close(fds[i]);
+}
+
+/* ------------------------------------------------------------------------ */
+
+/* Open all /dev/input/event* files. */
+static int
+open_fds(int fds[], int *num, int max_num)
+{
+	DIR *dir;
+	struct dirent *d;
+
+	if (!(dir = opendir(DEV_INPUT_DIR))) {
+		perrorf("Can't open directory %s", DEV_INPUT_DIR);
+		return -1;
+	}
+
+	while ((d = readdir(dir))) {
+		char name[256];
+
+		if (strncmp(d->d_name, EVENT_PREFIX, strlen(EVENT_PREFIX)))
+			continue; /* Not /dev/input/event* file */
+
+		snprintf(name, sizeof(name), "%s/%s", DEV_INPUT_DIR,
+			 d->d_name);
+		if ((fds[*num] = open(name, O_RDONLY)) == -1) {
+			perrorf("Can't open input device %s", name);
+			close_fds(fds, *num);
+			closedir(dir);
+			return -1;
+		}
+
+		if (++*num >= max_num)
+			break;
+	}
+
+	closedir(dir);
+	return 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+sysfs_write_int(const char *fname, int val)
+{
+	FILE *f;
+
+	if (!(f = fopen(fname, "w"))) {
+		perrorf("Can't open \"%s\" for writing", fname);
+		return -1;
+	}
+
+	fprintf(f, "%d\n", val);
+	fclose(f);
+	return 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+turn_display_on(void)
+{
+	return sysfs_write_int(DISPLAY_CONTROL, 0);
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+turn_display_off(void)
+{
+	return sysfs_write_int(DISPLAY_CONTROL, 1);
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+handle_events(int fd)
+{
+	char buf[BUF_SIZE];
```

Comment:
Even if the events are discarded, it might be clearer to use input_event buffers. Also it would implicitly make sure the i/o happens in multiple-of-event-size chunks.